metaseq/sequence_generator.py

```diff
@@ -280,6 +331,59 @@ def _generate(
             )
         return retval
 
+    def _initialize_generation_buffers(
+        self, bsz: int, beam_size: int, max_len: int, src_tokens: torch.Tensor
+    ) -> Tuple[
+        torch.Tensor, torch.Tensor, Optional[torch.Tensor], Optional[torch.Tensor]
+    ]:
+        """
+        Initialize buffers for use during generation.
+
+        Args:
+            bsz (int): batch size
+            beam_size (int): beam size
+            max_len (int): max gen length
+            src_tokens (tensor): incoming context
+
+        Returns:
+            scores: token score buffer [bsz*beam_size, max_len]
+            tokens: generation tokens [bsz*beam_size, max_len]
+            count_tokens: token count buffer for in-generation repetition penalty
+                [bsz*beam_size, vocab_size]
+            count_tokens_src: token count buffer for in-context repetition penalty
+                [bsz*beam_size, vocab_size]
+        """
+        scores = torch.zeros(bsz * beam_size, max_len).to(src_tokens).float()
+        tokens = (
+            torch.zeros(bsz * beam_size, max_len).to(src_tokens).long().fill_(self.pad)
+        )
+        count_tokens = None
+        count_tokens_src = None
+        if self.alpha_presence > 0 or self.alpha_frequency > 0:
+            count_tokens = torch.zeros(bsz * beam_size, self.vocab_size).to(src_tokens)
+        if self.alpha_presence_src > 0 or self.alpha_frequency_src > 0:
+            # histc requires floats
+            float_src_tokens = src_tokens.float()
+            if self.alpha_src_penalty_end_idx > 0 and not (bsz > 1):
+                # ignore this parameter if we're in a batch. sorry!
```

Comment:
right now with how generation objects are constructed, different batch items might have different `alpha_src_penalty_end_idx`, but only one of these will be propagated. the solutions are:

1) include this parameter in the batching logic (so that batch items have the same value, which is highly unlikely)
2) modify sequence generator construction/batch logic to "batchify" the incoming values so that this is passed as a list or tensor rather than a singular value
3) just ignore during batching

I chose option 3 due to simplicity, but let me know if the other two sound more reasonable to you

Reply:
why is it unlikely for us to batch similar things with alpha_src_penalty_end_idx? I don't understand; can you point out where it comes from?

Reply:
The `alpha_src_penalty_end_idx` cuts off the incoming context at a certain point when considering repetition penalties. Different incoming contexts may have different end indices. A simple example is someone wishing to penalize the context up to, but not including, the last sentence. The variable nature of incoming context lengths will lead to differing `alpha_src_penalty_end_idx`.